### PR TITLE
feat(frontend/recs): group rows by cell with savings range + collapse/expand (closes #225, closes #226)

### LINKED ISSUE #225: feat(frontend/recs): show savings range per cell + overall page-level range
## Background

After PR #195 fan-out, each rec cell `(provider, account, service, region, resource_type, engine)` shows up to 6 `(term, payment)` variants — each with its own monthly savings figure. The user has to eyeball the 6 numbers per cell to understand the spread. There's no compact "for this resource, savings are between $X and $Y depending on which option you pick" view — and no overall "if you select cheapest options across the board you save $A; if you go max-savings you save $B" headline.

### Per-cell savings range

For each rec cell, compute and display:

```
min_monthly_savings = MIN over (term, payment) variants
max_monthly_savings = MAX over (term, payment) variants
```

Render in the table as a "Savings range" row at the cell-group header (depends on cell-grouping sibling issue), or — if rows stay flat — as a separate column on each row showing the cell-level min/max ("$45 / $80" or "$45–$80").

Same for the Effective savings % column (sibling #221) — `min_pct` / `max_pct` over the cell's variants.

### Overall range

On the rec-page summary banner (currently shows `total_monthly_savings` and `total_upfront_cost` headlines):

```
overall_min_savings  = sum over cells of MIN(savings) variant
overall_max_savings  = sum over cells of MAX(savings) variant
overall_min_upfront  = sum over cells of MIN(upfront_cost) variant — typically the no-upfront variant
overall_max_upfront  = sum over cells of MAX(upfront_cost) variant — typically the all-upfront variant
```

Rendered as bracket text below the headline currency:

> Total monthly savings: **$1,234** _(range: $980 — $1,540 across cell variants)_
> Total upfront: **$23,450** _(range: $0 — $58,100)_

The bracket helps the user gauge "how much does my payment-option choice matter" at a glance, and informs whether they should go cheap (low upfront) or save-max (high upfront).

### Selected-vs-range

A bonus that falls out of the same data: the **selected** total against the cell-group range:

> You've picked options that save $1,100/mo — the max possible across these cells is $1,540/mo.

That's the most action-driving form: it tells the user how much they're leaving on the table by their current selections.

## Implementation sketch

- All math is local to the frontend; the data is already on each rec (`rec.savings`, `rec.upfront_cost`, `rec.monthly_cost`).
- Group recs by `cellKey` (the prefix from the radio-button issue — define once and reuse).
- Per-cell aggregations are pure functions: `aggCellRange(recs[]): { minSavings, maxSavings, minUpfront, maxUpfront }`.
- Overall aggregation sums the per-cell aggregations.

Tests pin the math for: a single cell with all 6 variants, a cell with only `1yr` variants (ranges still computed), and a cell with one variant (range collapses to a single value — render `$X` not `$X–$X`).

## Acceptance

- Each rec cell surfaces the savings range across its variants.
- The page summary surfaces the overall savings + upfront range.
- A "selected vs max" line tells the user how much room they have to improve their selection.
- Math is unit-tested against fixtures with known-min and known-max variants.

## Sequencing

Depends on the cell-grouping issue for the per-cell display (otherwise there's no anchor for "cell range"). The overall range can land independently — it just needs the `cellKey` grouping helper.

## References

- PR #195 / issue #188 (the fan-out)
- Sibling issues:
  - "Default-select per cell" — uses the same cellKey
  - "Radio-button mutual exclusion" — uses the same cellKey
  - "Group rec rows by cell in the table" — provides the visual anchor for per-cell range
  - #221 (Effective savings % column — feeds in to the per-cell percentage range)
- `frontend/src/recommendations.ts:113-117` (current summary banner — where the overall range rendering goes)
- `frontend/src/recommendations.ts:1147-1158` (current row template)

### LINKED ISSUE #226: feat(frontend/recs): group rec rows by physical-resource cell so variants are visually adjacent
## Background

Adjacent finding while exploring the rec-table for the radio-button + savings-range issues. Filing as its own issue because it's the visual anchor that makes those two changes legible.

After PR #195 fan-out, the rec table shows up to 6 rows per `(provider, account, service, region, resource_type, engine)` cell — one per `(term, payment)` variant. Today the rows are sorted by whichever column the user picked (Savings, Upfront, Region, etc.), so the 6 variants of a single cell are **visually scattered** across the table.

For users this means:

- Comparing "what's the spread for THIS m5.large?" is a manual scroll-and-eyeball exercise.
- The radio-button rule (sibling issue: at most one variant per cell selected) has no visual cue; a user can check `1yr all-upfront` for a cell and not realize the implicit deselection of `3yr no-upfront` happened on row 47 down the page.
- Per-cell savings range (sibling issue) has no anchor — the min/max numbers belong on a header row, not floating per-row.

## What needs to happen

Group the rec table by `cellKey` (the same `(provider, account, service, region, resource_type, engine)` prefix the radio-button + savings-range issues use). Render:

```
[ ] aws  acct-A  EC2  m5.large (Linux)  us-east-1   8 instances
       └─ 1yr / no-upfront     $0 upfront    $X/mo    $40/mo savings    33%
       (•) 1yr / partial         $X upfront    $Y/mo    $52/mo savings    41%   ← settings default
       ( ) 1yr / all-upfront    $XX upfront   $0/mo    $58/mo savings    44%
       ( ) 3yr / no-upfront     $0 upfront    $Z/mo    $66/mo savings    51%
       ( ) 3yr / partial          $XXX        $W/mo    $74/mo savings    57%
       ( ) 3yr / all-upfront    $XXXX         $0/mo    $80/mo savings    61%
       Cell range:  $40 — $80 monthly savings  ·  $0 — $XXXX upfront
```

Concretely:

- A **cell-header row** showing the resource identity once (provider · account · service · resource_type · region · count). This is the row the user reads to know *what* they're committing to.
- The **6 variant rows** are indented children of the header; only one can be selected (radio behavior from the sibling).
- The **cell-footer row** carries the per-cell savings range (sibling issue #C).
- Sorting: the **cell groups** sort against each other by the user's chosen column (using the cell's max savings, or the selected variant's value — designer's call). Inside a cell, variants always sort by `(term, payment)` in a fixed order (stable, predictable).
- Filtering: per-column filters apply to the cell as a whole — a cell with even one variant matching the filter renders.

### The "no fan-out" case

Some recs have only one variant (e.g., AWS Savings Plans, GCP CUDs that don't have a payment-option dimension). For cells with just one variant, render the row inline (no group header) — the visual hierarchy is overhead the user doesn't need.

### Implementation sketch

- `groupRecsByCell(recs[]) → Map<cellKey, {header, variants[]}>` — pure helper.
- Render: replace the flat row template at `recommendations.ts:1147-1158` with a grouped template. Each group: 1 header `<tr>` with a `colspan` summary cell, N variant `<tr>` rows, optional 1 footer `<tr>` with the range.
- Existing per-column sort and filter logic operates on the groups; a small adapter wraps the per-row sort key as a per-group sort key (max-of-variants for the active sortable column).

## Acceptance

- Variants of the same physical resource are visually adjacent under a single cell header.
- Cells with only one variant render flat (no group overhead).
- Sorting + filtering still work intuitively (sorts at the cell level, filters at the cell-passes-if-any-variant-matches level).
- Existing rec-table tests are updated to assert the grouped shape; new tests pin the single-variant edge case and the sort-stability inside a group.

## Severity / priority

p3 — purely visual; the radio-button rule (sibling) is what fixes correctness. Cell-grouping is the polish that makes the radio rule and the savings range readable. Land them as a sequence: radio first (correctness) → grouping (legibility) → savings-range (the metric that grouping makes useful).

## References

- PR #195 / issue #188 (the fan-out)
- Sibling issues: default-select, radio-button mutual exclusion, savings range (all of them benefit from the visual grouping)
- `frontend/src/recommendations.ts:1147-1158` (the row template to refactor)
- `frontend/src/recommendations.ts:130-160` (sort/filter registries — adapt for groups)

### PR DESCRIPTION
**Pull request by cristim**
## Summary

Closes #225 and #226. This PR merges both issues into a single feature since they're tightly coupled (cell grouping is the visual anchor that makes savings range legible).

- **Cell grouping (#226)**: Recommendation rows are now grouped by physical-resource cell `(provider|account|service|region|resource_type|engine)` — the same key as #224's radio enforcement. Multi-variant cells render as a collapsed summary row with a chevron; single-variant cells render flat (no overhead). Expand-all toggle added to the filter-status bar.
- **Savings range (#225)**: The per-cell summary row shows `$X – $Y/mo` savings range across all (term, payment) variants. Collapses to `$X` when all variants are uniform. When a variant is selected (radio from #224), the summary row shows that variant's value with `(+N variants)` instead of the range. A page-level range banner at the top of the table shows the aggregate across all visible cells.

## Design choices (UX call — flag anything to change in CR comments)

1. **Collapsed by default** — cleaner first impression. Most users only need to see the best option per resource, not all 6 variants on first load.
2. **Summary row layout** — provider + account + service in the first 4 columns; a single wide content cell (colspan) with the chevron, resource identity, and range. This keeps the identity readable without wrapping. The identity cell has `colspan=8` (the remaining columns after the 4 identity ones).
3. **Expand-All button location** — in the filter-status bar (above the table), not in the bottom sticky toolbar. Keeps it near the table context rather than near the purchase controls.
4. **Sort operates at the cell level** — cells sort by the selected variant's column value if one is selected; otherwise by the max value for numeric columns (e.g. max-savings). String columns are uniform within a cell by definition (they're the cell key).
5. **Single-variant cells render flat** — no group overhead (no summary row, no chevron, no indentation). Per #226's acceptance criteria.
6. **CSS approach** — variant rows use `padding-left: 2rem` on all cells except the checkbox column. Summary rows have a slightly distinct background (`#f3f6fb`). No new CSS framework or component library.
7. **`css-modules.d.ts` addition** — fixes a pre-existing TypeScript 5.9 strict-mode error on the `import './styles.css'` side-effect import in `index.ts`. The base branch had this error but CI passed because of bundler resolution; adding the declaration makes local `tsc --noEmit` clean as well.

## Out of scope

- Effective savings % column (#221) — separate issue, separate PR.
- Sparkline (#239) — separate issue.
- Column-filter primitives (#166) — separate issue.
- Savings-vs-selected delta ("You've picked options that save $X; max possible is $Y") — listed as a bonus in #225. Deferred — the per-cell range data is now available to implement this in a follow-up.

## Test plan

- [ ] All 1452 existing tests still pass (`npm test`)
- [ ] TypeScript `--noEmit` clean
- [ ] `npm run build` compiles without errors
- [ ] 16 new tests added:
  - `groupRecsByCell` pure helper: 2 tests
  - `cellSummary` pure helper: 3 tests
  - `pageLevelRange` pure helper: 2 tests
  - DOM integration: collapsed-by-default, chevron expand, page banner, single-variant flat, expand-all button, expand-all functional: 6 tests
  - Existing #224 radio enforcement test updated to expand the cell first (collapsed-by-default means checkboxes aren't in the DOM until expanded)

🤖 Generated with [claude-flow](https://github.com/ruvnet/claude-flow)

## Release Notes

* **New Features**
  * Recommendations table now groups variants by physical cell with collapsible summary rows
  * Added "Expand all / Collapse all" toggle to manage multi-variant cell visibility
  * Displays "Recommended range" banner for pages with multiple variant options

### CODE REVIEW

**Comment by cristim** [comment]
@coderabbitai review

**Comment by cristim** [comment]
## CodeRabbit review — round 2 response

Addressed both actionable findings from the round 1 review:

**Finding 1 — provider injection into CSS class (security)**
- Added `providerBadgeClass()` helper that whitelists `rec.provider` against `aws|azure|gcp` before injecting into class attributes. Non-whitelisted values fall back to `''` (no badge class).
- Badge display text is now `escapeHtml(providerDisplayName(rec.provider))` in both the summary row and variant row markup paths.

**Finding 2 — cell purchase target regression**
- `resolvePurchaseTarget()` now uses `pickBestVariantPerCell(visible)` for the default (no-selection) path. This ensures the purchase target is exactly one rec per physical resource — matching what the user sees in the collapsed grouped table. Submitting all visible variants of a multi-variant cell would create conflicting reservations for the same resource, violating the one-variant-per-cell contract from #224.
- `updateBottomActionBox()` derives `visibleCellCount` from `pickBestVariantPerCell` for the default path and shows "N cells visible" to match the cell-grouped table count.

**Test updates**
- #111 (a)/(b)/(c)/(d) fixtures: changed to use distinct `resource_type` values per rec so each rec occupies its own cell. With `pickBestVariantPerCell` for the default target, same-cell recs (same account + same resource type) now collapse to a single target — tests that expected fan-out needed different cells.
- Bundle B summary text assertions updated to match the new "cells visible" format.

All 1452 tests pass, TypeScript clean, build clean.

@coderabbitai review

File: frontend/src/recommendations.ts [hunk 1]
```diff
@@ -1278,19 +1462,125 @@ const FILTERABLE_COLUMNS: readonly state.RecommendationsColumnId[] = [
...
+    <td><span class="service-badge">${escapeHtml(rec.service)}</span></td>
+    <td title="${escapeHtml(rec.resource_type)}">${escapeHtml(rec.resource_type)}${rec.engine ? ` (${escapeHtml(rec.engine)})` : ''}${badge}</td>
+    <td>${escapeHtml(rec.region)}</td>
+    <td>${rec.count}</td>
+    <td>${formatTerm(rec.term)}</td>
+    <td class="savings">${formatCurrency(rec.savings)}</td>
+    <td>${formatCurrency(rec.upfront_cost)}</td>
+    <td>${formatCurrency(rec.monthly_cost ?? 0)}</td>
+    <td${pctClass}>${pctText}</td>
+  </tr>`;
```
[the remark this answers is not in the document]
Comment (cristim) [reply frontend/src/recommendations.ts:1512]: Fixed in commit e22afd6fa5. Added `providerBadgeClass()` that whitelists provider against aws|azure|gcp and falls back to ''. Display text now uses `escapeHtml(providerDisplayName(rec.provider))`. Applied in both `buildVariantRowMarkup` and `buildListMarkup` (summary row).

File: frontend/src/recommendations.ts [hunk 2]
```diff
@@ -1278,19 +1462,125 @@ const FILTERABLE_COLUMNS: readonly state.RecommendationsColumnId[] = [
...
+
+  // Update module-level group key cache for the Expand-All button.
+  // Only multi-variant cells count — single-variant cells render flat with no chevron.
+  lastVisibleGroupKeys = sortedKeys.filter((k) => (groups.get(k)?.length ?? 0) > 1);
+
+  // Page-level range banner (summed across all visible cells).
+  const plr = pageLevelRange(groups);
+  const bannerHtml = plr.cellCount > 0 && plr.savingsMax > 0
+    ? `<div class="rec-range-banner" role="status" aria-live="polite">Recommended range: <strong>${formatSavingsRange(plr.savingsMin, plr.savingsMax)}/mo</strong> across ${plr.cellCount} cell${plr.cellCount === 1 ? '' : 's'}</div>`
+    : '';
```
[the remark this answers is not in the document]
Comment (cristim) [reply frontend/src/recommendations.ts:1541]: Fixed in commit e22afd6fa5. `resolvePurchaseTarget()` now uses `pickBestVariantPerCell(visible)` for the default (no-selection) path — one rec per physical cell. `updateBottomActionBox()` derives `visibleCellCount` from `pickBestVariantPerCell` and displays 'N cells visible' to match the grouped table view.

**Comment by cristim** [comment]
Rebased onto current `feat/multicloud-web-frontend` (which now includes PR #254's monthly_cost fix — null-aware rendering, NaN for numeric filters, POSITIVE_INFINITY for sort). Resolved conflicts in `frontend/src/recommendations.ts`: kept #253's cell-grouping structure (`rows.join('')`, `buildVariantRowMarkup`, expand/collapse, `groupRecsByCell`, `cellSummary`, `pageLevelRange`, `groupsInSortOrder`) + adopted #254's corrected null-aware `monthly_cost` data path (`monthly_cost != null ? formatCurrency(...) : '—'` in `buildVariantRowMarkup`; `?? Number.POSITIVE_INFINITY` in sort; `?? Number.NaN` in filter). TypeScript clean, 140 tests pass, webpack build clean. @coderabbitai review
